abi.html

```diff
@@ -5526,16 +5524,148 @@ <h5><a href="#mangle.template-arg">5.1.5.10 Template Arguments</a></h5>
+                 ::= &lt;<a href="#mangle.template-param-decl">template-param-decl</a>&gt; &lt;<a href="#mangle.template-arg">template-arg</a>&gt;       # converted template argument
 </code></font></pre>
 
 <p>
-Type arguments appear using their regular encoding.
-For example, the template class "A&lt;char, float&gt;" is encoded as "1AIcfE".
-A slightly more involved example is
-a dependent function parameter type "A&lt;T2&gt;::X"
-(T2 is the second template parameter)
-which is encoded as "N1AIT0_E1XE",
-where the "N...E" construct is used to describe a qualified name.
+Template arguments can be dependent when a template argument list
+is written in a context where dependent structures must be mangled.
+For example, if the parameter type of a function template is
+<code>A&lt;T2&gt;::X</code>, where <code>T2</code> is the second
+template parameter of the function template, this is mangled as
+<code class=mangle>N1AIT0_E1XE</code>.
+
+<p>
+The mangling of template arguments in dependent contexts depends on
+whether the template argument can be matched with a template parameter
+and, if so, whether the argument can be checked against the parameter.
+In general, mangling assumes that all possible matching and checking
+has been performed.  The conditions are not precisely laid out here,
+but matching is generally inhibited by an unresolved template name
+and in some situations with unexpanded pack expansions, while checking
+is generally inhibited by dependent arguments or parameters.
+In non-dependent contexts (in particular, in a top-level
+<code>&lt;encoding&gt;</code>), template arguments should always be
+fully matched and checked against the template parameters of the
+specialized template.
+
+<p>
+Under the ODR, function templates are distinguished by differences
+in their template head, such as a constraint on a template type
+parameter, the type of a non-type template parameter, or (recursively)
+the template head of a template template parameter.  The template
+parameter list is not directly included in the
+<code>&lt;encoding&gt;</code> of a function template specialization
+(except for the <code>operator()</code> of a generic lambda),
+but it is indirectly reflected in the mangling of the template
+arguments as follows.  A template parameter is said to be <i>natural</i>
+for a template argument that has been matched and checked against
+it if:
+<ul>
+<li>the argument is an empty pack and the parameter is an
+  unconstrained template type parameter pack;
+<li>the argument is a non-empty pack and a non-pack variant
+  of the parameter would be natural for the first element of
+  the pack;
+<li>the argument is a template and the parameter has the exact
+  same template head;
+<li>the argument is a type and the parameter is unconstrained; or
+<li>the argument is a non-type template argument and the declared
+  parameter type neither is instantiation dependent nor contains
```

Comment:
is it intended to change the mangling of C++98 code like
```
template <class T, T V> T f() { return V; }
int main() { return f<int,42>(); }
```
most of the templates I see that use `<typename T, T V>' are classes and so unaffected, but there's at least one internal helper function in libstdc++ that will be changed.

Reply:
Saying that only a type parameter pack can be natural for an empty argument pack also changes the mangling of C++11 code like
```
template <int i> int max() { return i; }
template <int i, int j, int... rest> int max()
{ 
  int sub = max<j, rest...>();
  return i > sub ? i : sub;
}
int main() {  return max<1,2,3>(); }
```

and requiring template template args to have the same template head changes the mangling of C++11 code like
```
template <template <typename...> class TT, typename... Ts> TT<Ts...> f();
template <typename> struct A { }; 
int main() { f<A,int>(); }
```

Reply:
Some mangling changes for function templates are unavoidable, yeah.  Even in C++98, we have the conflict:

```c++
// _Z3fooIiLi0EEvv
template <class T, T N> void foo() {}
template void foo<int, 0>();

// _Z3fooIiLi0EEvv
template <class T, int N> void foo() {}
template void foo<int, 0>();
```

You can't have these both in the same translation unit, but they're formally different templates and need to be mangled differently.